proxy/src/serve_tasks.rs

```diff
@@ -178,13 +188,131 @@ async fn validate_and_remove_signatures(json: &mut Value) -> bool {
     true
 }
 
-async fn validate_helper_value<M: Msg + DeserializeOwned + std::fmt::Debug> (value: &Value) -> bool {
+async fn validate_helper_value<M: Msg + DeserializeOwned + std::fmt::Debug>(value: &Value) -> bool {
     let value = value.clone();
     match serde_json::from_value::<MsgSigned<M>>(value) {
         Ok(msg) => {
             debug!("Verifying reply {:?}", msg);
             msg.verify().await.is_ok()
-        },
+        }
         Err(_) => true, // Not of this type -> considered "valid"
     }
 }
+
+fn decryption_helper(value: &mut Value) -> Result<(), SamplyBeamError> {
+    if value.is_array() {
+        for inner in value.as_array_mut().unwrap() {
+            return decryption_helper(inner);
+        }
+    } else if value.is_object() {
+        if is_message_type::<EncryptedMsgTaskRequest>(value) {
+            *value = decrypt_msg::<MsgTaskRequest, EncryptedMsgTaskRequest>(value)?;
+            return Ok(());
+        } else if is_message_type::<EncryptedMsgTaskResult>(value) {
+            *value = decrypt_msg::<MsgTaskResult, EncryptedMsgTaskResult>(value)?;
+            return Ok(());
+        }
+    }
+    *value = value.clone();
+    Ok(())
+}
+
+fn is_message_type<M: Msg + DeserializeOwned + std::fmt::Debug>(value: &Value) -> bool {
```

Comment:
Trait on Msg

Reply:
This can be implemented very generically in the `Msg` trait, once [specialization](https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md) becomes stable (see [example](https://stackoverflow.com/a/60138532)).
At the moment, there is no generic way to test for type equality. However, a comment is added to the code for future refactoring.